docs/lesson-plans/week-5.md

```diff
@@ -0,0 +1,639 @@
+        ...
+        this.processor = new Processor(this.db, new TestExterns());
+        ...
+    }
+}
+```
+
+#### Exceptions
+
+When testing exceptions, JUnit provides methods and annotations to help, but we find it more helpful to use the 
+try/catch way of doing things. We think it allows you to test things about the exception better.
+
+Example:
+```java 
+try {
+    thingThatThrowsException();
+    fail(); // fail this test if we make it this far
+}
+catch (TheExpectedExceptionType e) {
+    assertEquals("My exception message", e.getMessage());
+    assertEquals(otherThingsAboutExceptionIfYouWant, e.doSomething());
+}
+```
+
+### Testing Using Mockito
+
+Mockito is a really useful library for mocking the interactions between objects that you don't have control over
+(like dependencies). It lets you create fake versions of objects which you have complete control over, and it even lets
+you verify things about how it was called.
+
+If you're interested in learning more about Mockito, [here is the 
+documentation](http://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html). It even lets you do some 
+really cool things like `spy()`, where you can pass in an existing object, have it do _most_ of the things it usually 
+does, and then just override/verify/mock what you want!
+
+#### Mocking an Object and Overriding Returns
+
+To mock an object, all you have to do is call (the static method) `mock(Thing.class)`, which will return a fake
+instance of that `Thing` class/interface. By default, it will return null 
+for every non-void method you call, but you can override that using `thenReturn`, `thenThrow`, `then` (`thenAnswer`, 
+allows you to do custom stuff), or `doReturn`, `doThrow`, `doAnswer`, `doNothing`, `doCallRealMethod` in the case of 
+void returns. You can also chain them together to have it return/throw/do things one after the other in the order 
+they're called.
+
+>An 'instance' of an object is a constructed value of the object created by its constructor. So for the class `Thing`, 
+>it's the difference between referring to `Thing.staticMethod()` as a static method and `new Thing().instanceMethod()`.
+
+Here are examples of the `then...` and `do...` styles of mocking:
+```java 
+MyCustomObject obj = mock(MyCustomObject.class); // can be done to interfaces too!
+
+when(obj.myCustomMethod()).thenReturn("Hello World!");
+
+String res = obj.myCustomMethod() // Returns "Hello World!");
+
+assertEquals("Hello World!", res);
+```
+
+```java 
+MyCustomObject obj = mock(MyCustomObject.class); // can be done to interfaces too!
+
+doReturn("Hello World!").when(obj).myCustomMethod());
+
+String res = obj.myCustomMethod() // Returns "Hello World!");
+
+assertEquals("Hello World!", res);
+```
+
+If you're interested in having the thing you want to do be done only when certain parameters are entered, then you can 
+do that with argument matchers. You can do things like `any()` for anything, `anyInt()` for any integer, 
```

Comment:
Mockito defines the matcher `any()` for anything, `anyInt()` for any integer, `any(MyClass.class)` for a specific class, or you can even just enter your own number/string for specific arguments.